src/sentry/sentry_apps/api/endpoints/installation_service_hook_projects.py

```diff
@@ -0,0 +1,191 @@
+                added_hook_projects
+                + list(ServiceHookProject.objects.filter(project_id__in=kept_project_ids)),
+                key=lambda x: x.project_id,
+            )
+
+    def _delete_servicehookprojects(self, service_hook_id: int, project_ids: set[int]) -> None:
+        ServiceHookProject.objects.filter(
+            service_hook_id=service_hook_id, project_id__in=project_ids
+        ).delete()
+
+    def _add_servicehookprojects(
+        self, service_hook_id: int, project_ids: set[int]
+    ) -> list[ServiceHookProject]:
+        res = []
+        for project_id in project_ids:
+            res.append(
+                ServiceHookProject.objects.create(
+                    project_id=project_id,
+                    service_hook_id=service_hook_id,
+                )
+            )
+        return res
+
+    def get(self, request: Request, installation: RpcSentryAppInstallation) -> Response:
+        hook = ServiceHook.objects.get(installation_id=installation.id)
+        hook_projects = list(ServiceHookProject.objects.filter(service_hook_id=hook.id))
+
+        for hp in hook_projects:
+            p_obj = project_service.get_by_id(
+                organization_id=installation.organization_id, id=hp.project_id
+            )
+            if not request.access.has_project_access(p_obj):
+                return Response(
+                    status=400,
+                    data={"error": "Some projects are not accessible"},
+                )
+
+        return self.paginate(
+            request=request,
+            queryset=hook_projects,
+            paginator_cls=OffsetPaginator,
+            on_results=lambda x: serialize(
+                x, request.user, access=request.access, serializer=ServiceHookProjectSerializer()
+            ),
+        )
+
+    def put(self, request: Request, installation: RpcSentryAppInstallation) -> Response:
+
+        serializer = ServiceHookProjectsInputSerializer(data=request.data)
+        if not serializer.is_valid():
+            return Response(serializer.errors, status=400)
+
+        projects = serializer.validated_data["projects"]
+
+        # convert slugs to ids if needed
+        org_id = installation.organization_id
+        project_ids = set()
+        for project in set(projects):
+            if isinstance(project, int):
+                project_obj = project_service.get_by_id(organization_id=org_id, id=project)
+            else:
+                project_obj = project_service.get_by_slug(organization_id=org_id, slug=project)
+            if project_obj and request.access.has_project_access(project_obj):
+                project_ids.add(project_obj.id)
+            else:
+                return Response(
+                    status=400,
+                    data={"error": f"Project '{project}' does not exist or is not accessible"},
+                )
+
+        try:
+            hook_projects = self._replace_hook_projects(installation, project_ids, request)
```

Comment:
You load all the projects in lines 148-154, and then again inside in `_replace_hook_projects`, could the project list be reused? Doing 2n queries could result in this endpoint having slower response times.

Reply:
those are 2 different non-overlapping sets of projects
besides this will be called once in a blue moon, I don't think it's worth optimizing